<a href="https://colab.research.google.com/github/Mohamad-Selawy/Data-Analysis/blob/main/Data%20Analysis/Pandas_Project_Solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🕵️‍♂️ Mission: Data Detective — The Curious Case of the Messy Superstore



---

### Welcome, Cadet! 🚀

A major Superstore Giant is in panic mode. Their systems are clogged, their databases are messy, and their stakeholders are asking questions they can't answer. You have been hired as the Lead Data Detective to audit their data, clean up the crime scenes, and extract business insights.

Are you ready to crack the case? Let's go!

## 📦 2. DATA INGESTION & MISSION SETUP

🚨 **Warning, Cadet!** The cargo (dataset) you're about to pull is incredibly messy and needs immediate inspection. Handle with care, and prepare for some serious data wrangling!

---


In [242]:
# 🚀 Run this cell to import your raw, unfiltered, messy cargo!
#!pip install kagglehub pandas
import kagglehub
import pandas as pd
import os

print("🚚 Fetching raw Superstore dataset...")
path = kagglehub.dataset_download("mohamadselawy/superstore-dataset/versions/4")
datafile = os.listdir(path)
df = pd.read_csv(f"{path}/{datafile[0]}") # Assuming the first file is the CSV
print("🎉 Cargo secured! Ready for the crime scene investigation.")

🚚 Fetching raw Superstore dataset...


100%|██████████| 760k/760k [00:00<00:00, 72.2MB/s]

Extracting files...
🎉 Cargo secured! Ready for the crime scene investigation.


## 🔍 3. STEP 1: CRIME SCENE INVESTIGATION (Data Auditing)

Alright, Detective! Your first mission: **Inspect the Clues.** The raw data is full of secrets and potential anomalies. Your job is to uncover them. Don't worry, the Superstore hasn't told us what's wrong – that's *your* job to find out!

Use the empty code cells below to write your investigative code. Look for anything suspicious!

### 📝 Detective's Audit Checklist:

*   **📐 Shape & Structure:** How big is this database? (Rows, Columns)
*   **🕳️ Black Holes:** Are there missing/null values hiding in key columns? If so, where and how many?
*   **👥 Clones:** Are there duplicate rows trying to skew our metrics?
*   **🔠 Sneaky Strings:** Are there trailing spaces, inconsistent casing, or weird, chaotic capitalization issues in text-based columns?
*   **🕰️ Typo-Datatypes:** Are dates stored as text? Are numbers pretending to be objects? Check if columns are in their correct data types.

---


### 📐 Shape & Structure

In [224]:
# Display the number of rows and columns
print(f"DataFrame Shape: {df.shape[0]} rows, {df.shape[1]} columns")

DataFrame Shape: 10382 rows, 21 columns


### 🕳️ Black Holes: Missing/Null Values

In [225]:
# Check for missing values
missing_values = df.isnull().sum()
missing_values = missing_values[missing_values > 0]
print("Missing Values per Column:")
display(missing_values)

Missing Values per Column:


,0
Ship Mode,308
Postal Code,515
Category,513
Sales,725
Profit,523


### 👥 Clones: Duplicate Rows

In [226]:
# Check for duplicate rows
duplicate_rows = df.duplicated().sum()
print(f"Number of Duplicate Rows: {duplicate_rows}")

Number of Duplicate Rows: 27


### 🕰️ Typo-Datatypes & General Info

In [227]:
# Display concise summary of the DataFrame, including data types and non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10382 entries, 0 to 10381
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         10382 non-null  int64  
 1   Order ID       10382 non-null  object 
 2   Order Date     10382 non-null  object 
 3   Ship Date      10382 non-null  object 
 4   Ship Mode      10074 non-null  object 
 5   Customer ID    10382 non-null  object 
 6   Customer Name  10382 non-null  object 
 7   Segment        10382 non-null  object 
 8   Country        10382 non-null  object 
 9   City           10382 non-null  object 
 10  State          10382 non-null  object 
 11  Postal Code    9867 non-null   float64
 12  Region         10382 non-null  object 
 13  Product ID     10382 non-null  object 
 14  Category       9869 non-null   object 
 15  Sub-Category   10382 non-null  object 
 16  Product Name   10382 non-null  object 
 17  Sales          9657 non-null   object 
 18  Quanti

### 🔠 Sneaky Strings: Inconsistent Casing, Trailing Spaces

In [240]:
# Inspect unique values for potential string inconsistencies in key columns
string_columns_to_check = ['Ship Mode', 'Segment', 'Country', 'Region', 'State', 'City', 'Category', 'Sub-Category', 'Product Name']

for col in string_columns_to_check:
      display(df[col].value_counts(dropna=False).sort_index())
      print(f"\n\n")
      if df[col].astype(str).apply(lambda x: (x.startswith(' ') or x.endswith(' '))).any():
          print(f"--- ALERT: Column '{col}' contains leading/trailing spaces ---")

,count
Ship Mode,
First Class,1561
Same Day,549
Second Class,1969
Standard Class,5995
NaN,308


,count
Segment,
CONSUMER,1119
CORPORATE,668
Consumer,3179
Corporate,1856
HOME OFFICE,375
Home Office,1153
consumer,1081
corporate,623
home office,328


,count
Country,
United States,10382


,count
Region,
Central,2411
East,2961
South,1696
West,3314


,count
State,
Alabama,66
Arizona,233
Arkansas,62
California,2064
Colorado,191
Connecticut,87
Delaware,100
District of Columbia,10
Florida,402


,count
City,
Aberdeen,1
Abilene,1
Akron,21
Albuquerque,14
Alexandria,17
...,...
Woonsocket,4
Yonkers,16
York,5


,count
Category,
Furniture,1796
Furnitures,77
Offcie Supplies,280
Office Supplies,5070
Tech,77
Technology,1582
furniture,207
office supplies,601
technology,179


,count
Sub-Category,
Accessories,806
Appliances,492
Art,839
Binders,1577
Bookcases,238
Chairs,644
Copiers,69
Envelopes,261
Fasteners,226


,count
Product Name,
"""WHILE YOU WERE OUT"" MESSAGE BOOK, ONE FORM PER PAGE\n",1
"""WHILE YOU WERE OUT"" MESSAGE BOOK, ONE FORM PER PAGE",1
"#10 GUMMED FLAP WHITE ENVELOPES, 100/BOX\n",1
#10 SELF-SEAL WHITE ENVELOPES,1
"#10 WHITE BUSINESS ENVELOPES,4 1/8 X 9 1/2",2
...,...
iOttie HLCRIO102 Car Mount,3
iOttie XL Car Mount,1
invisibleSHIELD by ZAGG Smudge-Free Screen Protector,6





--- ALERT: Column 'Product Name' contains leading/trailing spaces ---


## 🗒️ Detective's Notebook: Log Your Findings Here

**Findings from Step 1 (Crime Scene Investigation):**
*   **Anomaly 1:** `Sales` (725), `Profit` (523), `Postal Code` (515), `Category` (513), and `Ship Mode` (308) columns contain missing/null values.
*   **Anomaly 2:** There are `27` duplicate rows in the dataset.
*   **Anomaly 3:** `Order Date` and `Ship Date` are currently stored as `object` (string) types and need to be converted to `datetime` objects. `Sales` is also an `object` type and should be converted to a numeric type.
*   **Anomaly 4:** The `Postal Code` column, although a `float64`, contains missing values, which might indicate a need for specific handling or imputation.
*   **Anomaly 5:** `Ship Mode` contains `nan` values, confirming missing entries. Potential for leading/trailing spaces and inconsistent casing in other string columns (e.g., 'Category') should be addressed during cleaning.

**Actions Taken in Step 3 (The Deep Cleanse) on `df_2015`:**
*   **Dropped Duplicates:** `23` duplicate rows were successfully removed from the `df_2015` DataFrame.
*   **Data Type Conversion:** 'Order Date' and 'Ship Date' were converted to `datetime` format. 'Sales' and 'Profit' were cleaned (removed '$' and ',') and converted to `float64`.
*   **Missing Value Handling:** Missing 'Sales' and 'Profit' values were filled with `0`. Missing 'Ship Mode' and 'Category' values were filled with 'Unknown'. The 'Postal Code' column was removed during the pruning in Step 2, so its missing values are no longer a concern.
*   **String Normalization:** Key string columns ('Ship Mode', 'Segment', 'State', 'Category') had leading/trailing spaces stripped and casing normalized to title case.

## ✂️ 4. STEP 2: MISSION PREPARATION — PACK LIGHT (Subset Selection & Column Pruning)

**Golden Rule of Data Wrangling: Efficiency!** ⚡ As a seasoned detective, you never carry unnecessary luggage. We are **ONLY** interested in solving mysteries for the **year 2015**. Any other years are just background noise.

### The Pruning Rule:

Why clean data you'll never use? Identify which columns are strictly required to answer the 5 final business questions. Filter for orders from **2015 only**, keep only those essential columns, and discard the rest. Save your memory and computing power!

**Goal:** Create a new DataFrame containing *only* 2015 data and *only* the columns you deem necessary for the upcoming analysis.

---


In [229]:
# First, ensure 'Order Date' is in datetime format to filter by year
df['Order Date'] = df['Order Date'].astype('datetime64[ns]')

# Filter for orders from the year 2015
df_2015 = df[df['Order Date'].dt.year == 2015].copy()

# Define essential columns for the 5 business questions
essential_columns = [
    'Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
    'Segment', 'State', 'Sales', 'Profit', 'Category', 'Discount'
]

# Select only the essential columns for df_2015
df_2015 = df_2015[essential_columns]

print(f"Original DataFrame shape: {df.shape}")
print(f"2015 DataFrame shape (after filtering and column pruning): {df_2015.shape}")
display(df_2015.head())

Original DataFrame shape: (10382, 21)
2015 DataFrame shape (after filtering and column pruning): (2186, 11)


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Segment,State,Sales,Profit,Category,Discount
7,34,US-2015-150630,2015-09-17,9/21/2015,Standard Class,Consumer,Pennsylvania,NaN,3.5460,Office Supplies,0.2
8,305,CA-2015-144253,2015-05-04,5/9/2015,Second Class,corporate,New York,NaN,12.8640,Furniture,0.0
21,40,CA-2015-117415,2015-12-27,12/31/2015,Standard Class,Home Office,Texas,NaN,-15.1470,Furniture,0.3
25,5841,CA-2015-134992,2015-09-22,9/26/2015,Second Class,Home Office,Virginia,47.98,11.0354,NaN,0.0
30,3405,CA-2015-152527,2015-10-17,10/17/2015,NaN,consumer,California,77.88,3.8940,Office Supplies,0.0


## 🧹 5. STEP 3: THE DEEP CLEANSE (Data Cleaning Pipeline)

Now for the gritty work, Detective! It's time for **The Deep Cleanse**. Using your pruned 2015 subset, resolve all the anomalies you logged in your Detective's Notebook from Step 1.

**Your Cleaning Checklist:**

*   Strip any sneaky leading/trailing spaces from string columns.
*   Fix inconsistent capitalization where appropriate.
*   Convert date columns to proper datetime formats.
*   Cast numerical columns (like 'Sales' or 'Profit') to appropriate numeric types (e.g., floats).
*   Strategically tackle any remaining null values (e.g., fill with a sensible value, or drop if necessary).
*   Vaporize any duplicate rows that survived the pruning.

Ensure your data is pristine and ready for analysis!

---


In [230]:
# 1. Drop duplicate rows
initial_rows = df_2015.shape[0]
df_2015.drop_duplicates(inplace=True)
print(f"Dropped {initial_rows - df_2015.shape[0]} duplicate rows.")

# 2. Convert 'Ship Date' to datetime (Order Date was already converted in the previous step)
df_2015['Ship Date'] = df_2015['Ship Date'].astype('datetime64[ns]')

# 3. Clean and convert 'Sales' and 'Profit' to numeric
# Sales might contain non-numeric characters like '$' or commas, so clean first
df_2015['Sales'] = df_2015['Sales'].astype(str).str.replace('$', '').str.replace(',', '')
df_2015['Sales'] = df_2015['Sales'].astype('float64')
df_2015['Profit'] = df_2015['Profit'].astype('float64')

# Fill missing values for Sales and Profit with 0, as missing values likely mean no sale/profit.
df_2015['Sales'].fillna(0, inplace=True)
df_2015['Profit'].fillna(0, inplace=True)

# 4. Fill missing values for categorical columns 'Ship Mode' and 'Category'
df_2015['Ship Mode'].fillna('Unknown', inplace=True)
df_2015['Category'].fillna('Unknown', inplace=True)

# 5. Strip whitespace and normalize casing for string columns
string_cols = ['Ship Mode', 'Segment', 'State', 'Category']
for col in string_cols:
    if col in df_2015.columns:
        df_2015[col] = df_2015[col].astype(str).str.strip().str.title()

print("\n--- Cleaning Summary ---")
print("Data types after cleaning:")
display(df_2015.info())
print("\nMissing values after cleaning:")
display(df_2015.isnull().sum())
print("\nFirst 5 rows of cleaned data:")
display(df_2015.head())

Dropped 23 duplicate rows.

--- Cleaning Summary ---
Data types after cleaning:
<class 'pandas.core.frame.DataFrame'>
Index: 2163 entries, 7 to 10375
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Row ID      2163 non-null   int64         
 1   Order ID    2163 non-null   object        
 2   Order Date  2163 non-null   datetime64[ns]
 3   Ship Date   2163 non-null   datetime64[ns]
 4   Ship Mode   2163 non-null   object        
 5   Segment     2163 non-null   object        
 6   State       2163 non-null   object        
 7   Sales       2163 non-null   float64       
 8   Profit      2163 non-null   float64       
 9   Category    2163 non-null   object        
 10  Discount    2163 non-null   float64       
dtypes: datetime64[ns](2), float64(3), int64(1), object(5)
memory usage: 202.8+ KB


/tmp/ipykernel_1988/3296066127.py:16: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_2015['Sales'].fillna(0, inplace=True)
/tmp/ipykernel_1988/3296066127.py:17: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 

None


Missing values after cleaning:


,0
Row ID,0
Order ID,0
Order Date,0
Ship Date,0
Ship Mode,0
Segment,0
State,0
Sales,0
Profit,0
Category,0



First 5 rows of cleaned data:


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Segment,State,Sales,Profit,Category,Discount
7,34,US-2015-150630,2015-09-17,2015-09-21,Standard Class,Consumer,Pennsylvania,0.00,3.5460,Office Supplies,0.2
8,305,CA-2015-144253,2015-05-04,2015-05-09,Second Class,Corporate,New York,0.00,12.8640,Furniture,0.0
21,40,CA-2015-117415,2015-12-27,2015-12-31,Standard Class,Home Office,Texas,0.00,-15.1470,Furniture,0.3
25,5841,CA-2015-134992,2015-09-22,2015-09-26,Second Class,Home Office,Virginia,47.98,11.0354,Unknown,0.0
30,3405,CA-2015-152527,2015-10-17,2015-10-17,Unknown,Consumer,California,77.88,3.8940,Office Supplies,0.0


## 📈 6. STEP 4: ANSWERING THE BIG BUSINESS QUESTIONS

Great work, Detective! Your data is now sparkling clean and ready to spill its secrets. It's time to answer the Superstore's most pressing business questions. For each question below, write the code to find the answer and then articulate your **💡 Detective's Business Insight:** in the Markdown cell provided. What do the numbers truly mean for the Superstore?

---


### 📊 Question 1 (The Baseline): What are the total sales and total profit generated by the Superstore in 2015?


In [232]:
# Calculate total sales and total profit for 2015
total_sales_2015 = df_2015['Sales'].sum()
total_profit_2015 = df_2015['Profit'].sum()

print(f"Total Sales in 2015: ${total_sales_2015}")
print(f"Total Profit in 2015: ${total_profit_2015}")

Total Sales in 2015: $8447913.0215
Total Profit in 2015: $60446.0911


### 👥 Question 2 (The Target Market): Which customer segment (Consumer, Corporate, or Home Office) was the most profitable in 2015?


In [233]:
# Calculate profit by customer segment
profit_by_segment = df_2015.groupby('Segment')['Profit'].sum().sort_values(ascending=False)

print("Profit by Customer Segment in 2015:")
display(profit_by_segment)

Profit by Customer Segment in 2015:


,Profit
Segment,
Consumer,27783.7156
Corporate,19953.5315
Home Office,12708.8440


#### 💡 Detective's Business Insight:
In 2015, the **Consumer** segment was the most profitable, generating **$27,783.72** in profit. This indicates that individual customers are a key driver of the Superstore's bottom line, followed by Corporate and then Home Office segments. This insight can help in tailoring marketing strategies and resource allocation to further nurture the most profitable customer base.

### 🗺️ Question 3 (The Hotspots): What are the top 5 states by total Sales in 2015, and what were their corresponding profits?


In [234]:
# Calculate total sales and profit by state, then sort by sales
state_performance = df_2015.groupby('State').agg(
    {'Sales': 'sum',
    'Profit': 'sum'}
).sort_values(by = 'Sales', ascending=False)

# Get the top 5 states by sales
top_5_states = state_performance.head(5)

print("Top 5 States by Total Sales in 2015:")
display(top_5_states)

Top 5 States by Total Sales in 2015:


,Sales,Profit
State,,
New York,3077481.621,19334.8272
Washington,1023664.910,5337.9924
Ohio,1015240.854,-5196.7096
Georgia,1010898.250,4965.3701
Missouri,1006966.870,2892.6640


#### 💡 Detective's Business Insight:
The top 5 states by total sales in 2015 were New York, Washington, Ohio, Georgia, and Missouri. While New York leads significantly in both sales and profit, Ohio stands out as a state with high sales but a negative profit, indicating potential issues with pricing, discounts, or operational costs in that region. This suggests a need for further investigation into Ohio's market strategy.

### 🚚 Question 4 (The Logistics Race): Which shipping mode was most frequently utilized in 2015, and what was its average delivery time (Ship Date minus Order Date) in days?


In [235]:
# Calculate the delivery time for each order
df_2015['Delivery_Time'] = (df_2015['Ship Date'] - df_2015['Order Date']).dt.days

# Find the most frequently utilized shipping mode
most_frequent_ship_mode = df_2015['Ship Mode'].mode()[0]

# Calculate the average delivery time for the most frequent shipping mode
average_delivery_time = df_2015[df_2015['Ship Mode'] == most_frequent_ship_mode]['Delivery_Time'].mean()

print(f"Most Frequently Utilized Ship Mode: {most_frequent_ship_mode}")
print(f"Average Delivery Time for '{most_frequent_ship_mode}': {average_delivery_time:.2f} days")

Most Frequently Utilized Ship Mode: Standard Class
Average Delivery Time for 'Standard Class': 4.97 days


#### 💡 Detective's Business Insight:
In 2015, **Standard Class** was the most frequently utilized shipping mode, with an average delivery time of approximately **4.97 days**. This highlights its critical role in the Superstore's logistics. While efficient, understanding this benchmark allows for future optimization and comparison against other shipping options, ensuring competitive and satisfactory delivery experiences.

### 🎟️ Question 5 (The Discount Trap): What are the top product categories by sales in 2015, and what was the average discount offered for each?


In [236]:
# Calculate total sales and average discount by product category
category_performance = df_2015.groupby('Category').agg(
    Total_Sales=('Sales', 'sum'),
    Average_Discount=('Discount', 'mean')
).sort_values(by='Total_Sales', ascending=False)

print("Top Product Categories by Sales in 2015 and their Average Discounts:")
display(category_performance)

Top Product Categories by Sales in 2015 and their Average Discounts:


,Total_Sales,Average_Discount
Category,,
Office Supplies,6.117944e+06,0.162021
Technology,1.140988e+06,0.127056
Offcie Supplies,1.007591e+06,0.161667
Furniture,1.522621e+05,0.172009
Unknown,1.456669e+04,0.126786
Tech,1.001444e+04,0.133333
Furnitures,4.546347e+03,0.171765


#### 💡 Detective's Business Insight:
In 2015, **Office Supplies** led significantly in sales with **6,117,944**, followed by **Technology** with **$1,140,988**. The average discounts for these categories were 16.20% and 12.71% respectively. The presence of 'Offcie Supplies' and 'Furnitures' suggests further normalization of categorical data could be beneficial. This analysis highlights the importance of Office Supplies to the Superstore's revenue, but also raises questions about the profitability of discounts within this category, especially given the presence of similar categories with slightly different spellings and the 'Unknown' category.